File: 2020/13.py
import sys

# Note: periods in the schedule are prime numbers.
start_time = int(sys.stdin.readline())
schedule = [(i, int(word)) for (i, word) in enumerate(sys.stdin.readline().strip().split(',')) if word != 'x']

def Part1():
    min_time = best_period = None
    for _, period in schedule:
        if period is None:
            continue
        t = (start_time + period - 1) // period * period
        if min_time is None or t < min_time:
            min_time = t
            best_period = period
    return (min_time - start_time) * best_period

def Part2():
    t = 0
    m = 1
    for (i, p) in schedule:
        while (t + i) % p:
            t += m
        m *= p
    return t

print(Part1())
print(Part2())

File: 2020/haskell/13.hs
import Data.List
import Data.List.Split

parseSchedule :: String -> [Maybe Int]
parseSchedule = map parsePart . splitOn ","
    where
        parsePart "x" = Nothing
        parsePart s   = Just (read s)

solvePart1 :: Int -> [Maybe Int] -> Int
solvePart1 startTime periods = (min_time - startTime) * period
    where
        (min_time, period) = minimum [((startTime `divUp` p) * p, p) | Just p <- periods]

        -- Integer division, like `div`, but rounding up instead.
        divUp a b = (a + b - 1) `div` b

solvePart2 :: [Maybe Int] -> Integer
solvePart2 periods = t
    where
        (t, _) = foldl' update (0, 1) fullSchedule

        -- List of (index, period) pairs, which are the remainders and divisors.
        fullSchedule :: [(Integer, Integer)]
        fullSchedule = [(toInteger i, toInteger p) | (i, Just p) <- zip [0..] periods]

        -- Given start time t, multiplier m, remainder i and divisor p,
        -- calculates t' and m', so that m' = m * p and  t' = t + k * m has
        -- remainder i modulo p.
        update :: (Integer, Integer) -> (Integer, Integer) -> (Integer, Integer)
        update (t, m) (i, p) = (
            head $ dropWhile (\t -> (t + i) `mod` p /= 0) $ iterate (+ m) t,
            m * p)

main = do
    input <- getContents
    let [startTimeString, scheduleString] = lines input
    let startTime = read startTimeString
    let schedule = parseSchedule scheduleString
    print $ solvePart1 startTime schedule
    print $ solvePart2 schedule
